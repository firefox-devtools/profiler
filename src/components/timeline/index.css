/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
