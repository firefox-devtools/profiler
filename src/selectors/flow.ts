/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
