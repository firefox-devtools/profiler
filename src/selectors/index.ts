/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
