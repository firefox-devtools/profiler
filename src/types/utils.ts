/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
