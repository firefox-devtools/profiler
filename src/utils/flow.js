/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type { TabSlug } from '../app-logic/tabs-handling';
import type { TransformType } from 'firefox-profiler/types';

/**
 * This file contains utils that help Flow understand things better. Occasionally
 * statements can be logically equivalent, but Flow infers them in a specific way. Most
 * of the time tweaks can be done by editing the type system, but occasionally functions
 * are needed to get the desired result.
 */

/**
 * This function can be run as the default arm of a switch statement to ensure exhaustive
 * checking of a given type. It relies on an assumption that all cases will be handled
 * and the input to the function will be empty. This function hopefully makes that check
 * more readable.
 */
export function assertExhaustiveCheck(
  notValid: empty,
  errorMessage: string = `There was an unhandled case for the value: "${notValid}"`
): void {
  throw new Error(errorMessage);
}

/**
 * Immutably update an object through Object.assign, but retain the original
 * type information of the object. Flow will occasionally throw errors when
 * inferring what is going on with Object.assign.
 */
export function immutableUpdate<T>(object: T, ...rest: any[]): T {
  return Object.assign({}, object, ...rest);
}

/**
 * This function takes a string and returns either a valid TabSlug or null, this doesn't
 * throw an error so that any arbitrary string can be converted, e.g. from a URL.
 */
export function toValidTabSlug(tabSlug: any): TabSlug | null {
  const coercedTabSlug = (tabSlug: TabSlug);
  switch (coercedTabSlug) {
    case 'calltree':
    case 'function-list':
    case 'stack-chart':
    case 'marker-chart':
    case 'network-chart':
    case 'marker-table':
    case 'flame-graph':
    case 'js-tracer':
      return coercedTabSlug;
    default: {
      // The coerced type SHOULD be empty here. If in reality we get
      // here, then it's not a valid transform type, so return null.
      (coercedTabSlug: empty);
      return null;
    }
  }
}

/**
 * This function will take an arbitrary string, and will turn it into a TabSlug
 * it will throw an error if an invalid type was passed to it.
 */
export function ensureIsValidTabSlug(type: string): TabSlug {
  const assertedType = toValidTabSlug(type);
  if (!assertedType) {
    throw new Error(
      `Attempted to assert that "${type}" is a valid TransformType, and it was not.`
    );
  }
  return assertedType;
}

/**
 * This function will take an arbitrary string, and try to convert it to a valid
 * TransformType.
 */
export function convertToTransformType(type: string): TransformType | null {
  // Coerce this into a TransformType even if it's not one.
  const coercedType = ((type: any): TransformType);
  switch (coercedType) {
    // Exhaustively check each TransformType. The default arm will assert that
    // we have been exhaustive.
    case 'merge-call-node':
    case 'merge-function':
    case 'focus-subtree':
    case 'focus-function':
    case 'focus-category':
    case 'collapse-resource':
    case 'collapse-direct-recursion':
    case 'collapse-recursion':
    case 'collapse-function-subtree':
    case 'drop-function':
    case 'filter-samples':
      return coercedType;
    default: {
      // The coerced type SHOULD be empty here. If in reality we get
      // here, then it's not a valid transform type, so return null.
      (coercedType: empty);
      return null;
    }
  }
}

/**
 * This function coerces one type into another type.
 * This is equivalent to: (((value: A): any): B)
 */
export function coerce<A, B>(item: A): B {
  return (item: any);
}

/**
 * It can be helpful to coerce one type that matches the shape of another.
 */
export function coerceMatchingShape<T>(item: $Shape<T>): T {
  return (item: any);
}

/**
 * This is a type-friendly version of Object.values that assumes the object has
 * a Map-like structure.
 */
export function objectValues<Value, Obj: {| [string]: Value |}>(
  object: Obj
): Value[] {
  return (Object.values: any)(object);
}

/**
 * This is a type-friendly version of Object.entries that assumes the object has
 * a Map-like structure.
 */
export function objectEntries<Key, Value>(object: {
  [Key]: Value,
}): Array<[Key, Value]> {
  return (Object.entries: any)(object);
}

/**
 * This is a type-friendly version of Object.entries that assumes the object has
 * a Map-like structure.
 */
export function objectMap<Return, Key, Value>(
  object: { [Key]: Value },
  fn: (Value, Key) => Return
): { [Key]: Return } {
  const result: { [Key]: Return } = {};
  for (const [key, value] of objectEntries(object)) {
    result[key] = fn(value, key);
  }
  return result;
}

// Generic bounds with an Object is a false positive.
// eslint-disable-next-line flowtype/no-weak-types
export function getObjectValuesAsUnion<T: Object>(obj: T): Array<$Values<T>> {
  return Object.values(obj);
}

/**
 * This function will take an arbitrary string, and will turn it into a TransformType
 * it will throw an error if an invalid type was passed to it.
 */
export function ensureIsTransformType(type: string): TransformType {
  const assertedType = convertToTransformType(type);
  if (!assertedType) {
    throw new Error(
      `Attempted to assert that "${type}" is a valid TransformType, and it was not.`
    );
  }
  return assertedType;
}

export function ensureExists<T>(item: ?T, message: ?string): T {
  if (item === null) {
    throw new Error(message || 'Expected an item to exist, and it was null.');
  }
  if (item === undefined) {
    throw new Error(
      message || 'Expected an item to exist, and it was undefined.'
    );
  }
  return item;
}

/**
 * Returns the first item from Set in a type friendly manner.
 */
export function getFirstItemFromSet<T>(set: Set<T>): T | void {
  return set.values().next().value;
}
