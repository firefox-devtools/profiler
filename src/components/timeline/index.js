/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { FullTimeline } from 'firefox-profiler/components/timeline/FullTimeline';

type TimelineProps = {||};

export class Timeline extends PureComponent<TimelineProps> {
  // This may contain a function that's called whenever we want to remove the
  // "wheel" listener.
  _removeWheelListener: null | (() => mixed) = null;

  // This effectively disable the pinch-to-zoom as well as ctrl+mousewheel
  // gestures. Indeed in the timeline it is confusing. In the future we'll want
  // to couple this with the preview selection like in the Viewport HOC.
  preventPinchToZoom(e: WheelEvent) {
    if (e.ctrlKey) {
      e.preventDefault();
    }
  }

  // This will be registered as a ref property to the DOM element displaying the
  // tracks. We use this solution of registering the wheel event in this ref
  // listener because:
  // * we can't use React's event handling, because it doesn't allow us to use
  //   the "passive: false" way of registering the event handler. But we need this
  //   if we want to be able to prevent the default action of page zooming.
  // * we want to be sure to register it whenever the element changes.
  _onTimelineMountWithRef = (ref: HTMLElement | null) => {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }

    if (!ref) {
      return;
    }

    // without pinning to a const variable, Flow isn't sure that we don't change
    // the `ref` variable in some of the function calls below, and therefore
    // that it won't be null.
    const existingRef = ref;

    // Disable pinch-to-zoom and ctrl + wheel otherwise, on the timeline.
    // Indeed the users are used to this gesture  to zoom in in our charts, and
    // may use the same gesture elsewhere because of their habits, however this
    // doesn't do what they expect and instead zooms in the page, which is distracting.
    existingRef.addEventListener('wheel', this.preventPinchToZoom, {
      passive: false,
    });

    this._removeWheelListener = () => {
      existingRef.removeEventListener('wheel', this.preventPinchToZoom, {
        passive: false,
      });
    };
  };

  componentWillUnmount() {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }
  }

  render() {
    return <FullTimeline innerElementRef={this._onTimelineMountWithRef} />;
  }
}
