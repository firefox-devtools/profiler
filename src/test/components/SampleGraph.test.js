/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Provider } from 'react-redux';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { fireEvent } from '@testing-library/react';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import { ensureExists } from 'firefox-profiler/utils/flow';
import { TimelineTrackThread } from 'firefox-profiler/components/timeline/TrackThread';
import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  fireFullClick,
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';
import {
  autoMockElementSize,
  setMockedElementSize,
} from '../fixtures/mocks/element-size';
import {
  autoMockIntersectionObserver,
  triggerIntersectionObservers,
} from '../fixtures/mocks/intersection-observer';
import { triggerResizeObservers } from '../fixtures/mocks/resize-observer';

import type {
  Profile,
  IndexIntoSamplesTable,
  CssPixels,
} from 'firefox-profiler/types';

// Mocking the ActivityGraph because we don't want to see the content/draw log
// of it in these tests. It has its own tests.
jest.mock('firefox-profiler/components/shared/thread/ActivityGraph', () => ({
  ThreadActivityGraph: 'thread-activity-graph',
}));

// The following constants determine the size of the drawn graph.
const SAMPLE_COUNT = 8;
const PIXELS_PER_SAMPLE = 10;
const GRAPH_WIDTH = PIXELS_PER_SAMPLE * SAMPLE_COUNT;
const GRAPH_HEIGHT = 10;
function getSamplesPixelPosition(
  sampleIndex: IndexIntoSamplesTable
): CssPixels {
  // Compute the pixel position of the exact sample.
  return sampleIndex * PIXELS_PER_SAMPLE;
}

function getSamplesProfile() {
  return getProfileFromTextSamples(`
    A[cat:DOM]  A[cat:DOM]       A[cat:DOM]    A[cat:DOM]    A[cat:DOM]    A[cat:DOM]    A[cat:DOM]    A[cat:DOM]
    B           B                B             B             B             B             B             B
    C           C                H[cat:Other]  H[cat:Other]  H[cat:Other]  H[cat:Other]  H[cat:Other]  C
    D           F[cat:Graphics]  I             I             I             I             I             F[cat:Graphics]
    E           G                                                                                      G
  `).profile;
}

describe('SampleGraph', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  autoMockIntersectionObserver();
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  function setup(profile: Profile = getSamplesProfile()) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const flushRafCalls = mockRaf();

    const renderResult = render(
      <Provider store={store}>
        <TimelineTrackThread
          threadsKey={0}
          trackType="expanded"
          trackName="Test Track"
        />
      </Provider>
    );

    // WithSize uses requestAnimationFrame
    flushRafCalls();

    const sampleGraphCanvas = ensureExists(
      document.querySelector('.threadSampleGraphCanvas'),
      `Couldn't find the sample graph canvas, with selector .threadSampleGraphCanvas`
    );
    const thread = profile.threads[0];
    const { stringArray } = profile.shared;

    // Perform a click on the sample graph.
    function clickSampleGraph(index: IndexIntoSamplesTable) {
      fireFullClick(sampleGraphCanvas, {
        pageX: getSamplesPixelPosition(index),
        offsetX: getSamplesPixelPosition(index),
        pageY: GRAPH_HEIGHT / 2,
      });
    }

    // Hover over the sample graph.
    function hoverSampleGraph(index: IndexIntoSamplesTable) {
      fireEvent(
        sampleGraphCanvas,
        getMouseEvent('mousemove', {
          pageX: getSamplesPixelPosition(index),
          offsetX: getSamplesPixelPosition(index),
          pageY: GRAPH_HEIGHT / 2,
        })
      );
    }

    // This function gets the selected call node path as a list of function names.
    function getCallNodePath() {
      return selectedThreadSelectors
        .getSelectedCallNodePath(getState())
        .map((funcIndex) => stringArray[thread.funcTable.name[funcIndex]]);
    }

    /**
     * Coordinate the flushing of the requestAnimationFrame and the draw calls.
     */
    function getContextDrawCalls() {
      flushRafCalls();
      return flushDrawLog();
    }

    return {
      ...renderResult,
      dispatch,
      getState,
      profile,
      thread,
      store,
      sampleGraphCanvas,
      clickSampleGraph,
      hoverSampleGraph,
      getCallNodePath,
      getContextDrawCalls,
    };
  }

  it('matches the component snapshot', () => {
    const { sampleGraphCanvas } = setup();
    expect(sampleGraphCanvas).toMatchSnapshot();
  });

  it('matches the 2d canvas draw snapshot', () => {
    setup();
    expect(flushDrawLog()).toMatchSnapshot();
  });

  /**
   * The ThreadSampleGraph is not a connected component. It's easiest to test it
   * as once it's connected to the Redux store in the TrackThread.
   */
  describe('ThreadSampleGraph', function () {
    it('selects the full call node path when clicked', function () {
      const { clickSampleGraph, getCallNodePath } = setup();

      // The full call node at this sample is:
      //  A -> B -> C -> F -> G
      clickSampleGraph(1);
      expect(getCallNodePath()).toEqual(['A', 'B', 'C', 'F', 'G']);

      // The full call node at this sample is:
      //  A -> B -> H -> I
      clickSampleGraph(2);
      expect(getCallNodePath()).toEqual(['A', 'B', 'H', 'I']);
    });

    it('clicking outside of any sample removes any selection', function () {
      const { clickSampleGraph, getCallNodePath, sampleGraphCanvas } = setup();

      // Starting while nothing is selected.
      expect(getCallNodePath()).toEqual([]);

      // Selecting the sample with index 1.
      // The full call node at this sample is:
      //  A -> B -> C -> F -> G
      clickSampleGraph(1);
      expect(getCallNodePath()).toEqual(['A', 'B', 'C', 'F', 'G']);

      // Now we are selecting outside of the sample, which should remove the selection.
      fireFullClick(sampleGraphCanvas, {
        pageX: getSamplesPixelPosition(1) + PIXELS_PER_SAMPLE / 2,
        offsetX: getSamplesPixelPosition(1) + PIXELS_PER_SAMPLE / 2,
        pageY: GRAPH_HEIGHT / 2,
      });
      expect(getCallNodePath()).toEqual([]);
    });

    it('shows the correct tooltip when hovered', function () {
      const { hoverSampleGraph, getCallNodePath } = setup();

      // Hovering the sample with index 1.
      // The full call node at this sample is:
      //  A -> B -> C -> F -> G
      hoverSampleGraph(1);

      // We didn't click, so selection should not change in the selected node path.
      expect(getCallNodePath()).toEqual([]);

      // Make sure that we have a tooltip.
      expect(
        ensureExists(
          document.querySelector('.tooltip'),
          'A tooltip component must exist for this test.'
        )
      ).toMatchSnapshot();
    });

    it('does not show a tooltip when outside of a sample is hovered', function () {
      const { hoverSampleGraph, getCallNodePath, sampleGraphCanvas } = setup();

      // The full call node at this sample is:
      //  A -> B -> C -> F -> G

      hoverSampleGraph(1);
      // We didn't click, so selection should not change.
      expect(getCallNodePath()).toEqual([]);

      // Make sure that we have a tooltip.
      expect(document.querySelector('.tooltip')).toBeTruthy();

      // Now we are hovering outside of the samples.
      fireEvent(
        sampleGraphCanvas,
        getMouseEvent('mousemove', {
          pageX: getSamplesPixelPosition(1) + PIXELS_PER_SAMPLE / 2,
          offsetX: getSamplesPixelPosition(1) + PIXELS_PER_SAMPLE / 2,
          pageY: GRAPH_HEIGHT / 2,
        })
      );

      // There should be no tooltip this time
      expect(document.querySelector('.tooltip')).toBeFalsy();
    });
  });
});

describe('SampleGraph with intersection observer', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  // Do not automatically trigger the intersection observers.
  autoMockIntersectionObserver(false);

  function setup(profile: Profile = getSamplesProfile()) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const flushRafCalls = mockRaf();

    const renderResult = render(
      <Provider store={store}>
        <TimelineTrackThread
          threadsKey={0}
          trackType="expanded"
          trackName="Test Track"
        />
      </Provider>
    );

    // WithSize uses requestAnimationFrame
    flushRafCalls();

    /**
     * Coordinate the flushing of the requestAnimationFrame and the draw calls.
     */
    function getContextDrawCalls() {
      flushRafCalls();
      return flushDrawLog();
    }

    return {
      ...renderResult,
      dispatch,
      getState,
      profile,
      store,
      getContextDrawCalls,
    };
  }

  it('will not draw before the intersection observer', () => {
    const { getContextDrawCalls } = setup();
    const drawCalls = getContextDrawCalls();
    // There are other canvases inside the TrackThread too. We want to make sure
    // that sample graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );
  });

  it('will not draw after the intersection observer if it is not intersecting', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // There are other canvases inside the TrackThread too. We want to make sure
    // that sample graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'beginPath')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it still
    // doesn't draw it.
    triggerIntersectionObservers({ isIntersecting: false });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );
  });

  it('will draw after the intersection observer if it is intersecting', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // There are other canvases inside the TrackThread too. We want to make sure
    // that sample graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it draws it.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );
  });

  it('will redraw after it becomes visible again', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // There are other canvases inside the TrackThread too. We want to make sure
    // that activity graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it draws it.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );

    // Now it goes out of view again. Make sure that we don't redraw.
    triggerIntersectionObservers({ isIntersecting: false });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Send out the resize with a width change.
    // By changing the "fake" result of getBoundingClientRect, we ensure that
    // the pure components rerender because their `width` props change.
    setMockedElementSize({ width: GRAPH_WIDTH * 2, height: GRAPH_HEIGHT });
    triggerResizeObservers();
    drawCalls = getContextDrawCalls();
    // It should still be not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer again and make sure that it redraws.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );
  });
});
