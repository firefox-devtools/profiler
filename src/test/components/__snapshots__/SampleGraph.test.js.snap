// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleGraph ThreadSampleGraph shows the correct tooltip when hovered 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 21px; top: 10px;"
>
  <div
    class="tooltipDetails"
  >
    <div
      class="tooltipLabel"
    >
      Sampled at:
    </div>
    <div>
      1ms
    </div>
  </div>
  <div
    class="tooltipDetails"
  >
    <div
      class="tooltipLabel"
    >
      Category:
    </div>
    <div>
      <span
        class="colored-square category-color-green"
      />
      Graphics
    </div>
  </div>
  <div
    class="tooltipDetails"
  >
    <div
      class="tooltipLabel"
    >
      Stack:
    </div>
  </div>
  <ul
    class="backtrace"
  >
    <li
      class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
    >
      <span
        class="colored-border category-color-green"
        title="Graphics"
      />
      G
      <em
        class="backtraceStackFrameOrigin"
      />
    </li>
    <li
      class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
    >
      <span
        class="colored-border category-color-green"
        title="Graphics"
      />
      F
      <em
        class="backtraceStackFrameOrigin"
      />
    </li>
    <li
      class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
    >
      <span
        class="colored-border category-color-blue"
        title="DOM"
      />
      C
      <em
        class="backtraceStackFrameOrigin"
      />
    </li>
    <li
      class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
    >
      <span
        class="colored-border category-color-blue"
        title="DOM"
      />
      B
      <em
        class="backtraceStackFrameOrigin"
      />
    </li>
    <li
      class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
    >
      <span
        class="colored-border category-color-blue"
        title="DOM"
      />
      A
      <em
        class="backtraceStackFrameOrigin"
      />
    </li>
  </ul>
</div>
`;

exports[`SampleGraph matches the 2d canvas draw snapshot 1`] = `
Array [
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "fillRect",
    -0.4,
    0,
    0.8,
    0,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "fillRect",
    -5,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    5,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    15,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    25,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    35,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    45,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    55,
    0,
    10,
    10,
  ],
  Array [
    "fillRect",
    65,
    0,
    10,
    10,
  ],
  Array [
    "clearRect",
    0,
    0,
    80,
    10,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;

exports[`SampleGraph matches the component snapshot 1`] = `
<canvas
  class="threadSampleGraphCanvas threadSampleGraphCanvas"
  height="10"
  width="80"
>
  <h2>
    Stack Graph for 
    Test Track
  </h2>
  <p>
    This graph charts the stack height of each sample.
  </p>
</canvas>
`;
